chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,181 @@
+    of which element [sender, receiver] is the send/recv count for
+    send()/recv() call.
+
+    i.e.
+    sender rank i sends N elements to receiver j where N = table[i, j]
+    """
+
+    comm_size = len(length_all)
+    table = np.zeros([comm_size, comm_size], dtype=int)
+
+    # Distribute `length_all` elements over all processes
+    # To achieve a good balance, the reminder is scattered over
+    # processes of [rank, rank+1, ..., 0, 1]
+    for send_rank in range(comm_size):
+        sender_length = length_all[send_rank]
+        q = sender_length // comm_size
+        rem = sender_length % comm_size
+        for recv_rank in range(comm_size):
+            if (recv_rank - send_rank + comm_size) % comm_size < rem:
+                table[send_rank, recv_rank] = q + 1
+            else:
+                table[send_rank, recv_rank] = q
+        assert sum(table[send_rank, :]) == sender_length
+    return table
+
+
+def _send_recv_pairs(size):
+    pairs = [(x, y) for x in range(size) for y in range(size)]
+    # Sort the list, so send-recv communications from lower rank to higher rank
+    # happen earlier to avoid deadlock
+    # ex.)  [(0, 0), (1, 1), (0, 1), (1, 0)]
+    #     --> [(0, 0), (0, 1), (1, 0), (1, 1)]
+    pairs.sort()
+    return pairs
+
+
+def _exclusive_scan(array, dtype=None):
+    """C++'s std::exclusive_scan"""
+    if dtype is None:
+        dtype = array.dtype
+    z = np.zeros(len(array), dtype=dtype)
+    z[1:] = np.cumsum(array, dtype=dtype)[:-1]
+
+    return z
+
+
+def _exchange_chunks(comm, chunk, chunk_length_all):
+    count_table = _count_table(chunk_length_all)
+
+    data = []
+
+    # Perform parallel communication (which is very similar to alltoallv)
+    # using send() and recv()
+
+    # Generate all (sender, receiver) pairs
+    pairs = _send_recv_pairs(comm.size)
+
+    # offset of sendcounts
+    offset = _exclusive_scan(count_table[comm.rank, :])
+
+    # Perform send() and recv() in the order of `pairs`
+    # i.e.
+    #   pairs (0 --> 1) and (2 --> 3) are independent, so their
+    #   communications happen simultaneously
+    for send_rank, recv_rank in pairs:
+        count = count_table[send_rank, recv_rank]
+
+        if count == 0:
+            continue
+
+        if send_rank == recv_rank:
+            if send_rank == comm.rank:
+                # self copy
+                beg = offset[recv_rank]
+                end = beg + count
+                data += chunk[beg:end]
+        elif send_rank == comm.rank:
+            assert recv_rank != comm.rank
+            # I am the sender
+            beg = offset[recv_rank]
+            end = beg + count
+            comm.send_obj(chunk[beg:end], dest=recv_rank, tag=0)
+        elif recv_rank == comm.rank:
+            # I am the receiver
+            assert send_rank != comm.rank
+            recv_data = comm.recv_obj(source=send_rank, tag=0)
+            data += recv_data
+
+    return data
```

Comment:
The behaviour of process rank `i > 0` is:

- (0, i) -> recv_obj from `0`
- (1, i) -> recv_obj from `1`
- ...
- (i, 0) -> send to `0`
- (i, 1) -> send to `1`
- ...
- (i, i) -> copy internally
- ...
- (i, n-1) -> send to `n-1`
- ...
- (n-2, i) -> recv_obj from rank `n-2`
- (n-1, i) -> recv_obj from rank `n-1`

Which mean, at the first turn all process other than 0 is waiting. I don't think data transfer doesn't happen in parallel. Maybe this can be optimized by replacing the double-nested loop of send/recv with single loop `scatter_obj()` ( I know it's not yet implemented), for at least code readability? It also can leverage mpi4py's implementation.

Reply:
I think the algorithm is completely refactored.